src/op_mode/serial.py

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env python3
+#
+# Copyright (C) 2024 VyOS maintainers and contributors
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 2 or later as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import sys, os, re, json
+
+import vyos.opmode
+from vyos.base import Warning
+from vyos.utils.io import ask_yes_no
+from vyos.utils.process import cmd
+
+GLOB_GETTY_UNITS = 'serial-getty@*.service'
+RE_GETTY_DEVICES = re.compile(r'.+@(.+).service$')
+
+SD_UNIT_PATH = '/run/systemd/system'
+UTMP_PATH = '/run/utmp'
+
+def _get_serial_units():
+    # Since we cannot depend on the current config for decommissioned ports, 
+    # we just grab everything that systemd knows about. 
+    tmp = cmd(f'systemctl list-units {GLOB_GETTY_UNITS} --all --output json --no-pager')
+    getty_units = json.loads(tmp)
+    for sdunit in getty_units:
+        m = RE_GETTY_DEVICES.search(sdunit['unit'])
+        if m is None:
+            Warning(f'Serial console unit name "{sdunit["unit"]}" is malformed and cannot be checked for activity!')
+        else:
+            sdunit['device'] = m.group(1)
+
+    return getty_units
+
+def _get_connected_ports(units):
+    connected = []
+
+    ports = [ x['device'] for x in units if 'device' in x ]
+    for line in cmd(f'utmpdump {UTMP_PATH}').splitlines():
+        row = line.split('] [')
+        user_name = row[3].strip()
+        user_term = row[4].strip()
+        if user_name and user_name != 'LOGIN' and user_term in ports:
+            connected.append(user_term)
+
+    return connected
+
+def restart_getty(no_prompt: bool):
```

Comment:
First, I don't think we should mention "getty" in the function name. The user doesn't care what application provides the service, the logical entity that the user is working with is serial console.

Second, should we limit the user to only being able to restart _all_ consoles at once? I believe the function should provide a way to restart only specific consoles. I'm not sure what the best arguments would be: we can either make "all" the default if `console=None`, or add two mutually-exclusive arguments like `all: bool` and `console: str`. Making "all" the default is simpler and probably enough.

Reply:
1. I get your point, but it's not user-facing or at least, the users are developers. It's an internal implementation with a name only the op-mode schema or direct callers needs to worry about. I changed it from something like restart_serial_console to be more precise about what was happening. It's not much effort to change it back to match the user-facing command names. 
2. I did think about this and ended up removing it from the original flat version of the op-mode script as not relevant. It used a regex pattern argument defaulting to a match-all. The issue will be how to handle interaction with the conf_mode portion - that case would be broken if the restarts/stops weren't able to consider deleted devices. Perhaps separate pattern args along the lines of restart_consoles and cleanup_consoles so it's explicit to the caller, but only restart_consoles is exposed to op-mode usage (if at all).